src/android/toga_android/widgets/base.py

```diff
@@ -1,10 +1,15 @@
+from ..libs.activity import MainActivity
+
 
 class Widget:
     def __init__(self, interface):
         self.interface = interface
         self.interface._impl = self
         self._container = None
         self.native = None
+        # Capture a reference to the Java `MainActivity` instance, so that subclasses
+        # can pass it as `context` when creating native Android widgets.
+        self._native_activity = MainActivity.singletonThis
```

Comment:
There's nothing inherently wrong with this; however, in terms of data flow, I wonder if it might make sense for this to be set in set_app. In that context, you're explicitly being given the app impl, from which `app.native` would be the activity (or could be - android.App isn't tracking native at present, AFAICT).

Reply:
I struggled a little with this. I'm open to other ideas.

One difficulty with using `set_app()` is that it seems, from my manual testing, that it's called after a `Widget` (in this case, `Button`)'s constructor is called. I'm hoping to create the native Android widget from `Button.create()`, so I need the `MainActivity.singletonThis` somehow, and I think it needs to be before `set_app()` is called.

Reply:
Right - and this sort of data flow is why there's a separation between constructor and create. 

In practice, I recognize there's not a *lot* of difference; it's mostly just bugs me that we have a global singleton throughout the entire codebase.

Reply:
I have two options as I see it. I don't love of any of them :), but given the choices, I prefer this singleton.

Option 1. Current behavior. In the `Widget` class, grab this singleton. (Upside: Easy. Downside: Ugly.) This allows me to create the native widget in `create()`, called from the constructor.

Option 2. Adjust `create` to **not** create the widget, waiting for the `set_app()` call. Upside: Avoids singleton. Downside: The constructor wouldn't really create the widget. Instead, the constructor would store all the creation state, and when the user calls `set_app()` call, which happens when the widget is added to another widget, I'd create the widget. Furthermore, `set_label()` and other property writers would also have to write to Python state, since `self.native` wouldn't exist at the time that `set_label()` runs, since `create()` hasn't been called. This means I'd have a patch like this, which I've tested, and it works.

```
commit ef9e95a13c15945ea7db33dae99168c64cd5c282 (HEAD -> add-android-buttton)
Author: Asheesh Laroia <asheesh@asheesh.org>
Date:   Tue Apr 28 19:16:54 2020 -0700

    Delay create() until set_app()

diff --git a/src/android/toga_android/app.py b/src/android/toga_android/app.py
index b00f742b..0490e1a4 100644
--- a/src/android/toga_android/app.py
+++ b/src/android/toga_android/app.py
@@ -37,12 +37,12 @@ class TogaApp(IPythonApp):
         print("Toga app: onRestart")
 
 
-class App(IPythonApp):
+class App:
     def __init__(self, interface):
         self.interface = interface
         self.interface._impl = self
         self._listener = None
-        self.native = None
+        self.native = MainActivity.singletonThis
 
     def create(self):
         # The `_listener` listens for activity event callbacks. For simplicity,
diff --git a/src/android/toga_android/widgets/base.py b/src/android/toga_android/widgets/base.py
index ea276cb4..cdbce021 100644
--- a/src/android/toga_android/widgets/base.py
+++ b/src/android/toga_android/widgets/base.py
@@ -1,19 +1,26 @@
-from ..libs.activity import MainActivity
-
-
 class Widget:
     def __init__(self, interface):
         self.interface = interface
         self.interface._impl = self
         self._container = None
         self.native = None
-        # Capture a reference to the Java `MainActivity` instance, so that subclasses
-        # can pass it as `context` when creating native Android widgets.
-        self._native_activity = MainActivity.singletonThis
-        self.create()
+        self.app = None
+        # In Android, widgets need a reference to the app. Therefore, we call
+        # `create()` from `set_app()`, rather than from the constructor.
 
     def set_app(self, app):
-        pass
+        # I am hoping for the Android app. Check that app.native is defined.
+        if getattr(app, 'native', 'sentinel') != 'sentinel':
+            self.app = app
+            self.create()
+            return
+        # If it has '_impl', then grab that.
+        if getattr(app, '_impl', 'sentinel') != 'sentinel':
+            self.app = app._impl
+            self.create()
+            return
+        self.app = app
+        self.create()
 
     def set_window(self, window):
         pass
diff --git a/src/android/toga_android/widgets/box.py b/src/android/toga_android/widgets/box.py
index 661cf146..84b255e9 100644
--- a/src/android/toga_android/widgets/box.py
+++ b/src/android/toga_android/widgets/box.py
@@ -6,7 +6,23 @@ from ..libs.android_widgets import RelativeLayout
 class Box(Widget):
     def create(self):
         self.native = RelativeLayout(MainActivity.singletonThis)
+        children_queue = getattr(self, '_children_queue', None)
+        if children_queue is None:
+            return
+        for child in children_queue:
+            # If the child is lazily-created on Android, call set_app() to create it.
+            if child.native is None:
+                child.set_app(self.app)
+            super().add_child(child)
+            self.native.addView(child.native)
+        # Clear the queue
+        self._children_queue = None
 
     def add_child(self, child):
+        if self.native is None:
+            if getattr(self, '_children_queue', None) is None:
+                self._children_queue = []
+                self._children_queue.append(child)
+            return
         super().add_child(child)
         self.native.addView(child.native)
diff --git a/src/android/toga_android/widgets/button.py b/src/android/toga_android/widgets/button.py
index ad0c9605..e924485e 100644
--- a/src/android/toga_android/widgets/button.py
+++ b/src/android/toga_android/widgets/button.py
@@ -15,11 +15,20 @@ class TogaOnClickListener(android_widgets.OnClickListener):
 
 class Button(Widget):
     def create(self):
-        self.native = android_widgets.Button(self._native_activity)
+        self.native = android_widgets.Button(self.app.native)
+        queued_label = getattr(self, '_label', None)
+        if queued_label is not None:
+            self.native.setText(queued_label)
         self.native.setOnClickListener(TogaOnClickListener(button_impl=self))
 
     def set_label(self, label):
-        self.native.setText(self.interface.label)
+        if self.native:
+            self.native.setText(self.interface.label)
+        else:
+            # If `set_label()` is called before the widget is attached to a box
+            # or otherwise given its app, then `native` does not yet exist. Store
+            # the label as an instance property.
+            self._label = label
 
     def set_enabled(self, value):
         self.interface.factory.not_implemented('Button.set_enabled()')
@@ -32,6 +41,8 @@ class Button(Widget):
         pass
 
     def rehint(self):
+        if self.native is None:
+            return
         if self.native.getMeasuredWidth():
             # print("REHINT button", self, self.native.getMeasuredWidth(), self.native.getMeasuredHeight())
             self.interface.intrinsic.width = at_least(self.native.getMeasuredWidth() / self.app._impl.device_scale)
diff --git a/src/android/toga_android/window.py b/src/android/toga_android/window.py
index aa552fb5..991c7035 100644
--- a/src/android/toga_android/window.py
+++ b/src/android/toga_android/window.py
@@ -23,9 +23,26 @@ class Window:
         pass
 
     def set_app(self, app):
+        # I am hoping for the Android app. Check that app.native is defined.
+        if getattr(app, 'native', 'sentinel') != 'sentinel':
+            self.app = app
+            self.create()
+            return
+        # If it has '_impl', then grab that.
+        if getattr(app, '_impl', 'sentinel') != 'sentinel':
+            self.app = app._impl
+            self.create()
+            return
         self.app = app
+        self.create()
 
     def set_content(self, widget):
+        # If the widget is lazily-created on Android, create it now by calling set_app().
+        if widget.native is None:
+            widget.set_app(self.app)
+        if widget.native is None: # STILL???
+            print("set_content() is about to crash")
+            print(widget.__class__)
         self.app.native.setContentView(widget.native)
 
     def set_title(self, title):


```

I'm 100% some cleanup is possible, but the `if self.native` parts of that code are necessary, as far as I can tell. I feel the `if self.native` logic is not worth the trouble.

Penny for your thoughts.

Reply:
...oops, I didn't mean for that ` # STILL???` comment to come through. Oh well.

Reply:
Yeah - that's looking more messy than is worth it. As mentioned during our call, the lifecycle of creating widgets is something that may need further elaboration; in the meantime, accessing the singleton in widget base is fine.